File: src/content/algorithms/bubble-sort/code.py
def bubble_sort(array):  # #@function-def
    n = len(array)

    for i in range(n - 1):  # #@outer-loop
        swapped = False  # #@optimization-flag

        for j in range(n - i - 1):  # #@inner-loop
            if array[j] > array[j + 1]:  # #@comparison-swap
                array[j], array[j + 1] = array[j + 1], array[j]
                swapped = True

        if not swapped:  # #@early-exit
            break

    return array

File: src/content/algorithms/bubble-sort/code.rs
pub fn bubble_sort<T: Ord>(arr: &mut [T]) {  // #@generic-function
    let n = arr.len();

    for i in 0..n - 1 {  // #@outer-loop
        let mut swapped = false;

        for j in 0..n - i - 1 {  // #@inner-loop
            if arr[j] > arr[j + 1] {  // #@comparison-swap
                arr.swap(j, j + 1);  // #@swap-method
                swapped = true;
            }
        }

        if !swapped {  // #@early-exit
            break;
        }
    }
}
